include/tvm/tir/schedule/schedule.h

```diff
@@ -0,0 +1,208 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+#ifndef TVM_TIR_SCHEDULE_SCHEDULE_H_
+#define TVM_TIR_SCHEDULE_SCHEDULE_H_
+
+#include <tvm/tir/schedule/state.h>
+
+namespace tvm {
+namespace tir {
+
+/**************** Random variable: BlockRV ****************/
+
+/*! \brief A random variable that evaluates to a TensorIR block */
+class BlockRVNode : public runtime::Object {
+ public:
+  void VisitAttrs(tvm::AttrVisitor* v) {}
+  static constexpr const char* _type_key = "tir.BlockRV";
+  TVM_DECLARE_FINAL_OBJECT_INFO(BlockRVNode, runtime::Object);
+};
+
+/*!
+ * \brief Managed reference to BlockRVNode
+ * \sa BlockRVNode
+ */
+class BlockRV : public runtime::ObjectRef {
+ public:
+  /*! \brief Constructor */
+  TVM_DLL BlockRV();
+  TVM_DEFINE_NOTNULLABLE_OBJECT_REF_METHODS(BlockRV, runtime::ObjectRef, BlockRVNode);
+};
+
+/**************** Random variable: LoopRV ****************/
+
+/*! \brief A random variable that evaluates to a TensorIR for loop */
+class LoopRVNode : public runtime::Object {
+ public:
+  void VisitAttrs(tvm::AttrVisitor* v) {}
+  static constexpr const char* _type_key = "tir.LoopRV";
+  TVM_DECLARE_FINAL_OBJECT_INFO(LoopRVNode, runtime::Object);
+};
+
+/*!
+ * \brief Managed reference to LoopRVNode
+ * \sa LoopRVNode
+ */
+class LoopRV : public runtime::ObjectRef {
+ public:
+  /*! \brief Constructor */
+  TVM_DLL LoopRV();
+  TVM_DEFINE_NOTNULLABLE_OBJECT_REF_METHODS(LoopRV, runtime::ObjectRef, LoopRVNode);
+};
+
+/**************** Random variable: IntRV ****************/
+
+/*! \brief An integer random variable */
+using IntRV = PrimExpr;
+
+using IntRVNode = PrimExprNode;
+
+/**************** The schedule class ****************/
+
+class Schedule;
+
+/*!
+ * \brief The user-facing abstract schedule class
+ */
+class ScheduleNode : public runtime::Object {
+  friend class Schedule;
+
+ public:
+  virtual ~ScheduleNode() = default;
+
+  static constexpr const char* _type_key = "tir.Schedule";
+  TVM_DECLARE_BASE_OBJECT_INFO(ScheduleNode, runtime::Object);
+
+ public:
+  /*! \brief Take the IRModule out of the schedule */
+  virtual IRModule mod() const { return state()->mod; }
+  /*! \return The internal state of scheduling */
+  virtual ScheduleState state() const = 0;
+  /*!
+   * \brief Returns a copy of the schedule, including both the state and the symbol table,
+   * guaranteeing that
+   * 1) SRef tree is completely reconstructed;
+   * 2) The IRModule being scheduled is untouched;
+   * 3) All the random variables are valid in the copy, pointing to the correpsonding sref
```

Comment:
Can you clarify here? by pointing do you mean they all represent a corresponding sref?

Reply:
Yeah i find it somewhat hard to describe. Consider the code below:

```python
sch_1 = Schedule(...)
block_rv = sch_1.get_block(...)  # The block_rv points to a block sref in sch_1

# Note that the sref tree is deep copied
sch_2 = sch_1.copy()

# Then block_rv can point to different srefs in those schedules
block_sref_1 = sch_1.get_sref(block_rv)
block_sref_2 = sch_2.get_sref(block_rv)
assert not block_sref_1.same_as(block_sref_2)  # they are two different sref trees
```

Reply:
@jroesch would you like to help to rephrase :laughing:

Reply:
@tqchen @jroesch would you guys like to give me a hand in this docstring? Thanks in advance!

Reply:
the current docs looks good, let us followup with polishing PRs if needed

Reply:
The new description looks good to me, we can send followup PRs to improve